### PR TITLE
Fix tensorflow hanging test suite

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/12676?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12676/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12676
```

</p>
</details>

### Related Issues/PRs

Resolve failing CI https://github.com/mlflow-automation/mlflow/actions/runs/9940163012/job/27469161600

### What changes are proposed in this pull request?
The test hanging comes from `test_tf_keras_autolog_implicit_batch_size_works`.

Some threads are not terminated in the tests.  The fix is to always `flush_async_logging` after test ends, because `_batch_logging_worker_threadpool` and `_batch_status_check_threadpool` are only shutdown inside `flush`.
The tons of remaining `MLflowAsyncLoggingLoop` threads are expected referring to https://github.com/mlflow/mlflow/pull/10374.

Update:
root cause of leaking threads: 
https://github.com/mlflow/mlflow/blob/master/mlflow/tensorflow/__init__.py#L1343 uses flush_async_logging which sets up threads after terminating everything. In tensorflow autologging, we auto-patch fit functions, so within two `model.fit` call there're zombie threads left produced by `_set_up_logging_thread` call within `flush`. 
The fix is to use `shut_down_async_logging` in tensorflow autologging to always terminate the async logging. BTW `flush` should call `activate` instead of `_set_up_logging_thread` to make sure we set the status to ACTIVE, so the next time `activate` is called we don't reset the thread, we use the thread last `flush` creates, this prevents from threads shadowing (assign self._batch_logging_thread again while self._batch_logging_thread.is_alive() is True).
Since we don't know how many places involves log_batch_async, for test-only we always shutdown_async_logging to ensure no zombie threads left.

root cause of hanging: 
we didn't clear `_stop_data_logging_thread_event` status inside `end_async_logging`, so after this is called and another `_set_up_logging_thread` is called, we failed to enter `_log_run_data` inside `_logging_loop`, while `log_batch_async` keeps adding run_batch into the queue, so the job keeps hanging forever. 
The fix is to always clear `_stop_data_logging_thread_event` status inside `end_async_logging`.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/tensorflow/test_tensorflow2_autolog.py
Comment (harupy) [inline tests/tensorflow/test_tensorflow2_autolog.py:47]: `flush` sets up a new thread. It'd be nice if we could avoid setting up a new thread.

https://github.com/mlflow/mlflow/blob/6b4d9e7063cca7498132de966bb695ec77ae8584/mlflow/utils/async_logging/async_artifacts_logging_queue.py#L74

Comment (serena-ruan) [reply tests/tensorflow/test_tensorflow2_autolog.py:47]: Yes, I can add a separate method for this

File: mlflow/utils/async_logging/async_logging_queue.py
Comment (harupy) [inline mlflow/utils/async_logging/async_logging_queue.py:138]: can you elaborate on this?

Comment (serena-ruan) [reply mlflow/utils/async_logging/async_logging_queue.py:138]: from my test, this hangs (and it's not always hanging). From the documentation both self._queue.qsize and self._queue.empty() is not trustable, so I retry 3 times here to make sure the queue is actually empty.

Comment (harupy) [reply mlflow/utils/async_logging/async_logging_queue.py:138]: which line hangs?

Comment (serena-ruan) [reply mlflow/utils/async_logging/async_logging_queue.py:138]: I actually didn't go deeper, the test `test_tf_keras_autolog_implicit_batch_size_works` hangs, and I assume `log_batch_async` is the cause. I updated this and seems the test can consistently pass

Comment (harupy) [reply mlflow/utils/async_logging/async_logging_queue.py:138]: Instead of relying on `empty`, can we use `Queue.join`?

https://docs.python.org/3/library/queue.html#queue.Queue.join

Comment (serena-ruan) [reply mlflow/utils/async_logging/async_logging_queue.py]: Had another try, seems this is not relevant (can't repro the hang even without this change). But guess we can't use join because this is used for async_logging_buffer_seconds

File: mlflow/utils/async_logging/async_logging_queue.py
Comment (harupy) [inline mlflow/utils/async_logging/async_logging_queue.py:91]: I tried something similar but I wasn't able to clean up the zombie threads. Wondering what's missing.

File: mlflow/ml-package-versions.yml
Comment (harupy) [inline mlflow/ml-package-versions.yml:152]:
```suggestion
```

just a reminder to remove this

File: mlflow/__init__.py
Comment (harupy) [inline mlflow/__init__.py:225]:
```suggestion
    "_shut_down_async_logging",
```

I don't think we need to expose this as a public API.

File: mlflow/tracking/fluent.py
Comment (harupy) [inline mlflow/tracking/fluent.py:671]:
```suggestion
    """Shutdown all pending data and shutdown the async logging."""
```

File: tests/tensorflow/test_tensorflow2_autolog.py
Comment (harupy) [inline tests/tensorflow/test_tensorflow2_autolog.py:49]: We normally don't need try-finally in a pytest fixture. pytest handles errors.

Comment (serena-ruan) [reply tests/tensorflow/test_tensorflow2_autolog.py:49]: Ah ok!

**Comment by serena-ruan** [comment]
https://github.com/mlflow/mlflow/actions/runs/9974545441/job/27562276183?pr=12676 proves the fix works and no more zombie threads left

File: tests/utils/test_async_logging_queue.py
Comment (harupy) [inline tests/utils/test_async_logging_queue.py:76]: Can we use a pytest fixture?

Comment (serena-ruan) [reply tests/utils/test_async_logging_queue.py:76]: Ah yes, seems they all have the same initiation process

File: mlflow/tensorflow/__init__.py
Comment (harupy) [inline mlflow/tensorflow/__init__.py:1346]: @chenmoneygithub Curious what happens if we don't set up threads. Does the cause an issue?

Comment (serena-ruan) [reply mlflow/tensorflow/__init__.py:1346]: I think if we don't activate the async_logging but we use log_batch_async then there will be an issue

File: tests/utils/test_async_logging_queue.py
Comment (harupy) [inline tests/utils/test_async_logging_queue.py:92]: I'd use a context manger if I had to do this.

File: mlflow/utils/async_logging/async_logging_queue.py
Comment (harupy) [inline mlflow/utils/async_logging/async_logging_queue.py:301]: can we move this in the docstring?

**Review by harupy — APPROVED** [review]
Left a few more comments, LGTM once @chenmoneygithub also approves

File: tests/utils/test_async_logging_queue.py
Comment (chenmoneygithub) [reply tests/utils/test_async_logging_queue.py:92]: I agree with Haru, the current code is cool but kinda hard to read

**Review by chenmoneygithub — CHANGES_REQUESTED** [review]
Thanks for fixing, this is great work! Two things to call out:

- `qsize()` solution is required, because when we are reading the queue with time buffer, there are new data coming to the queue. `qsize()` controls how many to read to avoid the endless while loop.
- We are not activating the queue after `model.fit()`call, which would lead to a second `model.fit()` to fail to log the data. I read the PR description, but still don't quite understand why `active()` status leads to zombie thread.

**Review by chenmoneygithub — COMMENTED** [review]
Forgot one thing - we may also want to update this code https://github.com/mlflow/mlflow/blob/afd33b3d69e6da8a1d1b08112328d1f68d29604e/mlflow/tracking/_tracking_service/client.py#L888 to use `shut_down_async_logging` to reset the event.

File: mlflow/tracking/fluent.py
Comment (serena-ruan) [reply mlflow/tracking/fluent.py:671]: It's internal logic inside tensorflow though, we shouldn't expect customers to call it.
cc @harupy thoughts?

Comment (BenWilson2) [reply mlflow/tracking/fluent.py:671]: Leave it private. We don't want this to be exposed with a tab auto-complete for a public API as it is purely an internal API.

**Review by BenWilson2 — APPROVED** [review]
LGTM! Thanks for the fix and the great investigation here @serena-ruan and @harupy

File: mlflow/tracking/fluent.py
Comment (chenmoneygithub) [reply mlflow/tracking/fluent.py:671]: we can avoid exposing that at the top `mlflow` level, a method should be private if that's private to its own module. 

Not a big deal tho.

File: tests/tensorflow/test_tensorflow2_autolog.py
Comment (chenmoneygithub) [inline tests/tensorflow/test_tensorflow2_autolog.py:46]: curious on this - if we delete this line would the test fail? Basically users don't know that they need to call this method.

**Review by chenmoneygithub — APPROVED** [review]
LGTM!

File: tests/tensorflow/test_tensorflow2_autolog.py
Comment (serena-ruan) [reply tests/tensorflow/test_tensorflow2_autolog.py:46]: It doesn't fail, there are just some left threads

**Pull request merged**
